evm/trie/trie_test.go

```diff
@@ -0,0 +1,365 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package trie
+
+import (
+	"fmt"
+	"os"
+	"testing"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethdb/leveldb"
+	"github.com/ethereum/go-ethereum/rlp"
+)
+
+var (
+	defaultTriesToStore = 3
+)
+
+func createNewTxTries(numHistoricalTries int) *TxTries {
+	return NewTxTries(numHistoricalTries)
+}
+
+func createTempDB() *leveldb.Database {
+
+	diskdb, err := leveldb.New("./temp-database", 256, 0, "")
+	if err != nil {
+		panic(fmt.Sprintf("unable to create testing database: %v", err))
+	}
+	return diskdb
+}
+
+func deleteTempDB() {
+	os.RemoveAll("./temp-database")
+}
+
+func addTrie(txTries *TxTries, root common.Hash, transactions []common.Hash, db *leveldb.Database) error {
+
+	if db == nil {
+		db = createTempDB()
+	}
+
+	if transactions == nil {
+		transactions = make([]common.Hash, 0)
+	}
+
+	_, err := txTries.AddTrie(root, db, transactions)
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+
+}
+
+func TestEmptyTxTries(t *testing.T) {
+	txTries := createNewTxTries(defaultTriesToStore)
+
+	if txTries.triesToStore != defaultTriesToStore {
+		t.Fatalf("tries to store not set properly, expected: %x, got: %x", defaultTriesToStore, txTries.triesToStore)
+	}
+}
+
+func TestAddEmptyTrie(t *testing.T) {
+	txTries := createNewTxTries(defaultTriesToStore)
+	addTrie(txTries, emptyRoot, nil, nil)
+
+	if txTries.txRoots[0] != emptyRoot {
+		t.Fatalf("failed to set txRoot in txTries properly, expected: %x, got: %x", emptyRoot, txTries.txRoots[0])
+	}
+
+	if txTries.txTries[0].trie.Hash() != emptyRoot {
+		t.Fatalf("trie does not have empty hash as root, expected: %x, got: %x", emptyRoot, txTries.txTries[0].trie.Hash())
+	}
+
+	deleteTempDB()
+}
+
+func TestAddEmptyTrieRetrieveProof_Fails(t *testing.T) {
+	txTries := createNewTxTries(defaultTriesToStore)
+	addTrie(txTries, emptyRoot, nil, nil)
+
+	if txTries.txRoots[0] != emptyRoot {
+		t.Fatalf("failed to set txRoot in txTries properly, expected: %x, got: %x", emptyRoot, txTries.txRoots[0])
+	}
+
+	if txTries.txTries[0].trie.Hash() != emptyRoot {
+		t.Fatalf("trie does not have empty hash as root, expected: %x, got: %x", emptyRoot, txTries.txTries[0].trie.Hash())
+	}
+
+	deleteTempDB()
+}
+
+func TestAddSingleTrieUpdate(t *testing.T) {
+	vals := []common.Hash{common.HexToHash("123"), common.HexToHash("456")}
+	expectedRoot := common.HexToHash("8e4ba8a974b3dfa9dbfb76d61c633ba2a1250a4d79e3e912e841392fdb232c17")
```

Comment:
how are you computing these?

Reply:
inserting the values into the go-ethereum trie and getting the root. Would it be better if I set up a go-ethereum trie object as well and then compared the roots instead of having an expected root?

Reply:
You either need to document how they were generated or provide tooling to generate them again (eg. a script). Without this, no one but yourself could ever verify or update these.